numpy/core/src/umath/override.c

```diff
@@ -323,6 +323,94 @@ normalize_at_args(PyUFuncObject *ufunc, PyObject *args,
     return (*normal_args == NULL);
 }
 
+/*
+ * Check whether a set of input and output args have a non-default
+ *  `__array_ufunc__` method. Return the number of overrides, setting
+ * corresponding objects in PyObject array with_override and the corresponding
+ * __array_ufunc__ methods in methods (both using new references).
+ *
+ * Only the first override for a given class is returned.
+ *
+ * returns -1 on failure.
+ */
+NPY_NO_EXPORT int
+check_overrides(PyObject *args, PyObject *out_kwd_obj,
+                PyObject **with_override, PyObject **methods)
+{
+    int i;
+    PyObject *obj, *out_obj = NULL;
+
+    int num_override_args = 0;
+    int nin, nout;
+
+    nin = PyTuple_GET_SIZE(args);
+    if (out_kwd_obj == NULL) {
+        nout = 0;
+    }
+    else if (PyTuple_CheckExact(out_kwd_obj)) {
+        nout = PyTuple_GET_SIZE(out_kwd_obj);
+        if (nout == 1) {
+            /* saves special-casing the tuple-of-one below */
+            out_obj = PyTuple_GET_ITEM(out_kwd_obj, 0);
+        }
+    }
+    else {
+        nout = 1;
+        out_obj = out_kwd_obj;
+    }
+    for (i = 0; i < nin + nout; ++i) {
+        PyObject *method;
+        int j = 0;
+        npy_bool new_class = 1;
+
+        if (i < nin) {
+            obj = PyTuple_GET_ITEM(args, i);
+        }
+        else if (nout == 1) {
+            obj = out_obj;
+        }
+        else {
+            obj = PyTuple_GET_ITEM(out_kwd_obj, i - nin);
+        }
+        /*
+         * Now see if the object provides an __array_ufunc__. However, we should
+         * ignore the base ndarray.__ufunc__, so we skip any ndarray as well as
+         * any ndarray subclass instances that did not override __array_ufunc__.
+         *
+         * Furthermore, we should look at a given class only once.
+         */
+        while (j < num_override_args && new_class) {
+            new_class = (PyObject_Type(obj) != PyObject_Type(with_override[j]));
```

Comment:
Would it be better to store types in an array rather than calling `PyObject_Type` repeatedly?

Reply:
OOPS, this is the wrong function, I don't need a reference to the type at all, I just need `Py_TYPE`, which is very fast.

Reply:
I think this is wrong elsewhere in the function too...

Reply:
OK, I think I got them all.